File: QKSMS/src/main/java/com/moez/QKSMS/ui/base/QKFragment.java
package com.moez.QKSMS.ui.base;

import android.app.Activity;
import android.app.Fragment;
import android.os.Bundle;
import android.view.View;
import com.moez.QKSMS.QKSMSApp;
import com.moez.QKSMS.common.LiveViewManager;
import com.moez.QKSMS.enums.QKPreference;
import com.moez.QKSMS.ui.ThemeManager;
import com.squareup.leakcanary.RefWatcher;
import icepick.Icepick;

public class QKFragment extends Fragment {

    protected QKActivity mContext;

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        mContext = (QKActivity) activity;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        LiveViewManager.registerView(QKPreference.BACKGROUND, this, key -> {
            if (getView() != null) {
                getView().setBackgroundColor(ThemeManager.getBackgroundColor());
            }
        });
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Icepick.restoreInstanceState(this, savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        Icepick.saveInstanceState(this, outState);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        RefWatcher refWatcher = QKSMSApp.getRefWatcher(getActivity());
        refWatcher.watch(this);
    }
}

File: qksms/src/main/java/com/moez/QKSMS/ui/base/QkFragment.kt
package com.moez.QKSMS.ui.base

import android.arch.lifecycle.LifecycleFragment
import android.os.Bundle
import android.view.*
import com.moez.QKSMS.QKApplication
import com.moez.QKSMS.dagger.AppComponent

abstract class QkFragment : LifecycleFragment() {

    abstract fun provideLayoutRes(): Int
    open fun provideMenuRes(): Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(provideMenuRes() != 0)
    }

    final override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(provideLayoutRes(), container, false)
    }

    final override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        if (provideMenuRes() != 0) inflater.inflate(provideMenuRes(), menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun getContext(): QkActivity {
        return super.getContext() as QkActivity
    }

    fun getAppComponent(): AppComponent? {
        return (context.application as QKApplication).appComponent
    }

}